From 01bed8ba6e4ed90b1819b2be31658cddf1ae0c9f Mon Sep 17 00:00:00 2001
From: Tobias Schramm <t.schramm@t-sys.eu>
Date: Mon, 22 Mar 2021 00:43:38 +0100
Subject: [PATCH 09/11] ASoC: soc: codec: tas5825m: reconfigure gain control to
 linear scale

Previously the exposed gain settings were inverted and on a logarithmic
scale.

Signed-off-by: Tobias Schramm <t.schramm@t-sys.eu>
---
 sound/soc/codecs/tas5825m.c | 17 ++++++++++++-----
 1 file changed, 12 insertions(+), 5 deletions(-)

diff --git a/sound/soc/codecs/tas5825m.c b/sound/soc/codecs/tas5825m.c
index 67ac04b08599..ff4f13d04b41 100644
--- a/sound/soc/codecs/tas5825m.c
+++ b/sound/soc/codecs/tas5825m.c
@@ -353,7 +353,14 @@ static int tas5825m_codec_probe(struct snd_soc_component *component)
 				  TAS5825M_SDOUT_SEL_PREPROCESSING,
 				  TAS5825M_SDOUT_SEL_PREPROCESSING);
 	if (ret < 0)
-		goto error_snd_soc_component_update_bits;	
+		goto error_snd_soc_component_update_bits;
+
+	/* Set a low default volume */
+	ret = snd_soc_component_update_bits(component, TAS5825M_DIG_VOL,
+				  0xff,
+				  0xa0);
+	if (ret < 0)
+		goto error_snd_soc_component_update_bits;
 
 	/* Enter shutdown mode by default to minimize power consumption  */
 	ret = snd_soc_component_update_bits(component, TAS5825M_DEVICE_CTRL2,
@@ -545,15 +552,15 @@ static const struct regmap_config tas5825m_regmap_config = {
 	.volatile_reg = tas5825m_is_volatile_reg,
 };
 
-static const SNDRV_CTL_TLVD_DECLARE_DB_SCALE(pga_gain_tlv, SNDRV_CTL_TLVD_DB_SCALE_MUTE, 50, 2400);
+static const SNDRV_CTL_TLVD_DECLARE_DB_LINEAR(pga_gain_tlv, SNDRV_CTL_TLVD_DB_SCALE_MUTE, 2400);
 
-static const SNDRV_CTL_TLVD_DECLARE_DB_SCALE(again_tlv, -1550, 50, 0);
+static const SNDRV_CTL_TLVD_DECLARE_DB_LINEAR(again_tlv, -1550, 0);
 
 static const struct snd_kcontrol_new tas5825m_snd_controls[] = {
 	SOC_SINGLE_TLV("Speaker Driver Playback Volume",
-		       TAS5825M_DIG_VOL, 0, 0xff, 0, pga_gain_tlv),
+		       TAS5825M_DIG_VOL, 0, 0xff, 1, pga_gain_tlv),
 	SOC_SINGLE_TLV("Speaker Driver Analog Gain", TAS5825M_AGAIN,
-		       0, 0x1f, 0, again_tlv),
+		       0, 0x1f, 1, again_tlv),
 };
 
 static const struct snd_soc_dapm_widget tas5825m_dapm_widgets[] = {
-- 
2.30.1

